src/gui/properties/propertieswidget.cpp

```diff
@@ -663,78 +663,71 @@ void PropertiesWidget::displayWebSeedListMenu(const QPoint &)
-        if (!m_torrent || !m_torrent->hasMetadata()) return;
-        QString old_name = m_torrent->filePath(file_index);
-        if (old_name.endsWith(".!qB") && !new_name_last.endsWith(".!qB"))
-            new_name_last += ".!qB";
-        QStringList path_items = old_name.split("/");
-        path_items.removeLast();
-        path_items << new_name_last;
-        QString new_name = path_items.join("/");
-        if (old_name == new_name) {
-            qDebug("Name did not change");
+    if (PropListModel->itemType(modelIndex) == TorrentContentModelItem::FileType) {
+        // renaming a file
+        const int fileIndex = PropListModel->getFileIndex(modelIndex);
+
+        const QString oldFileName = m_torrent->fileName(fileIndex);
+        const QString oldFilePath = m_torrent->filePath(fileIndex);
+        const QString newFileName = (BitTorrent::Session::instance()->isAppendExtensionEnabled() && !newName.endsWith(QB_EXT))
+                                    ? newName + QB_EXT
+                                    : newName;
+        const QString newFilePath = oldFilePath.leftRef(oldFilePath.size() - oldFileName.size()) + newFileName;
+
+        if (oldFileName == newFileName) {
+            qDebug("Name did not change: %s", qPrintable(oldFileName));
             return;
         }
-        new_name = Utils::Fs::expandPath(new_name);
-        qDebug("New name: %s", qPrintable(new_name));
-        // Check if that name is already used
+
+        // check if that name is already used
         for (int i = 0; i < m_torrent->filesCount(); ++i) {
-            if (i == file_index) continue;
-            if (Utils::Fs::sameFileNames(m_torrent->filePath(i), new_name)) {
-                // Display error message
-                MessageBoxRaised::warning(this, tr("The file could not be renamed"),
+            if (i == fileIndex) continue;
+            if (Utils::Fs::sameFileNames(m_torrent->filePath(i), newFilePath)) {
+                MessageBoxRaised::warning(this, tr("Rename error"),
                                           tr("This name is already in use in this folder. Please use a different name."),
                                           QMessageBox::Ok);
                 return;
             }
         }
-        const bool force_recheck = QFile::exists(m_torrent->savePath(true) + "/" + new_name);
-        qDebug("Renaming %s to %s", qPrintable(old_name), qPrintable(new_name));
-        m_torrent->renameFile(file_index, new_name);
-        // Force recheck
-        if (force_recheck) m_torrent->forceRecheck();
-        // Rename if torrent files model too
-        if (new_name_last.endsWith(".!qB"))
-            new_name_last.chop(4);
-        PropListModel->setData(index, new_name_last);
+
+        qDebug("Renaming %s to %s", qPrintable(oldFilePath), qPrintable(newFilePath));
+        m_torrent->renameFile(fileIndex, newFilePath);
+
+        PropListModel->setData(modelIndex, newName);
```

Comment:
About the rename: The old code checks if the a file **on disk** exists with the new name. If it does, it schedules a recheck.
And personally, I think we are in uncharted territory in this case. If the renamed file hasn't been yet downloaded then the recheck probably will pickup the already existing file on disk(after the rename+recheck). But if the file has been created(old name) and there is already a file with the new name on disk what will happen after rename+recheck? Will libtorrent overwrite data by moving the file onto the exisiting one, will it fail the rename or will it use the existing one? Or is there a race condition that will do any of those 3 possibilities?